In [17]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [18]:
import numpy as np
import pandas as pd
# Reading csv file from drive using pandas library
df=pd.read_csv('/content/drive/MyDrive/PrepInsta/PrepInsta-DA-Week-3-main/Assignment1.csv')
df

Index   Age      Salary  Rating       Location  Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In         1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny         2002       TRUE
2       2   NaN   $77k-$89k    -1.0    New York,Ny           -1         -1
3       3  64.0   $44k-$99k     4.4       India In         1988         -1
4       4  25.0   $44k-$99k     6.4  Australia Aus         2002         -1
5       5  44.0   $77k-$89k     1.4       India,In         1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny           -1         -1
7       7  44.0   $44k-$99k    -1.0  Australia Aus           -1         -1
8       8  35.0   $44k-$99k     5.4    New York,Ny           -1         -1
9       9  22.0   $44k-$99k     7.7       India,In           -1       TRUE
10     10  55.0   $10k-$49k     5.4       India,In         2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In         2009         -1
12     12   NaN   $44k-$99k     0.0       India,In         1999         -1
13     13  25.0   $44k-$99k    -1.0  Australia Aus         2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus         2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus         1999         -1
16     16  19.0   $19k-$40k     4.5       India,In         1984         -1
17     17   NaN   $44k-$99k     5.3    New York,Ny         1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny         1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny         1955       TRUE
20     20   NaN   $44k-$99k     5.7    New York,Ny         1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny         1946         -1
22     22  19.0   $55k-$66k     7.8    New York,Ny         1988       TRUE
23     23   NaN   $44k-$99k     2.4    New York,Ny         1999       TRUE
24     24  13.0   $44k-$99k    -1.0    New York,Ny         1987         -1
25     25  55.0   $44k-$99k     0.0  Australia Aus         1980       TRUE
26     26   NaN   $55k-$66k     NaN       India,In         1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In         1935         -1
28     28   NaN   $39k-$88k     3.4  Australia Aus         1932         -1

Well, we will always try to remove the duplicates from a data frame for removing redundancies.

In [19]:
df=df.drop_duplicates()
df

Index   Age      Salary  Rating       Location  Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In         1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny         2002       TRUE
2       2   NaN   $77k-$89k    -1.0    New York,Ny           -1         -1
3       3  64.0   $44k-$99k     4.4       India In         1988         -1
4       4  25.0   $44k-$99k     6.4  Australia Aus         2002         -1
5       5  44.0   $77k-$89k     1.4       India,In         1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny           -1         -1
7       7  44.0   $44k-$99k    -1.0  Australia Aus           -1         -1
8       8  35.0   $44k-$99k     5.4    New York,Ny           -1         -1
9       9  22.0   $44k-$99k     7.7       India,In           -1       TRUE
10     10  55.0   $10k-$49k     5.4       India,In         2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In         2009         -1
12     12   NaN   $44k-$99k     0.0       India,In         1999         -1
13     13  25.0   $44k-$99k    -1.0  Australia Aus         2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus         2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus         1999         -1
16     16  19.0   $19k-$40k     4.5       India,In         1984         -1
17     17   NaN   $44k-$99k     5.3    New York,Ny         1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny         1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny         1955       TRUE
20     20   NaN   $44k-$99k     5.7    New York,Ny         1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny         1946         -1
22     22  19.0   $55k-$66k     7.8    New York,Ny         1988       TRUE
23     23   NaN   $44k-$99k     2.4    New York,Ny         1999       TRUE
24     24  13.0   $44k-$99k    -1.0    New York,Ny         1987         -1
25     25  55.0   $44k-$99k     0.0  Australia Aus         1980       TRUE
26     26   NaN   $55k-$66k     NaN       India,In         1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In         1935         -1
28     28   NaN   $39k-$88k     3.4  Australia Aus         1932         -1

`Easy Apply` column initially consists of string data type. Apart from that, it also consists of missing values which is `"-1"`. However, we consider `-1` as FALSE and replace all `-1` with `FALSE`.<br>
As the column's data type is still string, we will map respective values to respective boolean values.
<br>
After that we will replace all missing values, `'-1'` with `NaN`

In [20]:
df["Easy Apply"]=df["Easy Apply"].str.replace('-1','FALSE')
mapping = {'TRUE': True, 'FALSE': False}
df["Established"] = df["Established"].astype(str)
df = df.replace(-1, np.nan)
df

Index   Age      Salary  Rating       Location Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In        1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny        2002       TRUE
2       2   NaN   $77k-$89k     NaN    New York,Ny          -1      FALSE
3       3  64.0   $44k-$99k     4.4       India In        1988      FALSE
4       4  25.0   $44k-$99k     6.4  Australia Aus        2002      FALSE
5       5  44.0   $77k-$89k     1.4       India,In        1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny          -1      FALSE
7       7  44.0   $44k-$99k     NaN  Australia Aus          -1      FALSE
8       8  35.0   $44k-$99k     5.4    New York,Ny          -1      FALSE
9       9  22.0   $44k-$99k     7.7       India,In          -1       TRUE
10     10  55.0   $10k-$49k     5.4       India,In        2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In        2009      FALSE
12     12   NaN   $44k-$99k     0.0       India,In        1999      FALSE
13     13  25.0   $44k-$99k     NaN  Australia Aus        2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus        2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus        1999      FALSE
16     16  19.0   $19k-$40k     4.5       India,In        1984      FALSE
17     17   NaN   $44k-$99k     5.3    New York,Ny        1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny        1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny        1955       TRUE
20     20   NaN   $44k-$99k     5.7    New York,Ny        1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny        1946      FALSE
22     22  19.0   $55k-$66k     7.8    New York,Ny        1988       TRUE
23     23   NaN   $44k-$99k     2.4    New York,Ny        1999       TRUE
24     24  13.0   $44k-$99k     NaN    New York,Ny        1987      FALSE
25     25  55.0   $44k-$99k     0.0  Australia Aus        1980       TRUE
26     26   NaN   $55k-$66k     NaN       India,In        1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In        1935      FALSE
28     28   NaN   $39k-$88k     3.4  Australia Aus        1932      FALSE

As we observe that the data set size is small, dropping rows may occur large deviation of output. To keep the integrity, we will do the following steps:


1.   `Age:` Replace NaN values with the `.mean()` of Age column.
2.   `Rating:` Replace NaN values with the `.mean()` of Rating column.
3.   `Established:` Replace `'-1'` with the `"Unknown"` string.


In [21]:
age_mean=df["Age"].mean()
age_mean=int(age_mean)
df["Age"] = df["Age"].replace(np.nan, age_mean)

rating_mean=df["Rating"].mean()
rating_mean=int(rating_mean)
df["Rating"] = df["Rating"].replace(np.nan, rating_mean)


df["Established"] = df["Established"].replace('-1', "Unknown")
df

Index   Age      Salary  Rating       Location Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In        1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny        2002       TRUE
2       2  39.0   $77k-$89k     4.0    New York,Ny     Unknown      FALSE
3       3  64.0   $44k-$99k     4.4       India In        1988      FALSE
4       4  25.0   $44k-$99k     6.4  Australia Aus        2002      FALSE
5       5  44.0   $77k-$89k     1.4       India,In        1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny     Unknown      FALSE
7       7  44.0   $44k-$99k     4.0  Australia Aus     Unknown      FALSE
8       8  35.0   $44k-$99k     5.4    New York,Ny     Unknown      FALSE
9       9  22.0   $44k-$99k     7.7       India,In     Unknown       TRUE
10     10  55.0   $10k-$49k     5.4       India,In        2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In        2009      FALSE
12     12  39.0   $44k-$99k     0.0       India,In        1999      FALSE
13     13  25.0   $44k-$99k     4.0  Australia Aus        2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus        2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus        1999      FALSE
16     16  19.0   $19k-$40k     4.5       India,In        1984      FALSE
17     17  39.0   $44k-$99k     5.3    New York,Ny        1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny        1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny        1955       TRUE
20     20  39.0   $44k-$99k     5.7    New York,Ny        1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny        1946      FALSE
22     22  19.0   $55k-$66k     7.8    New York,Ny        1988       TRUE
23     23  39.0   $44k-$99k     2.4    New York,Ny        1999       TRUE
24     24  13.0   $44k-$99k     4.0    New York,Ny        1987      FALSE
25     25  55.0   $44k-$99k     0.0  Australia Aus        1980       TRUE
26     26  39.0   $55k-$66k     4.0       India,In        1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In        1935      FALSE
28     28  39.0   $39k-$88k     3.4  Australia Aus        1932      FALSE

Now drop all the rows, where we get all `NULL` / empty values.

In [22]:
df=df.dropna()
df

Index   Age      Salary  Rating       Location Established Easy Apply
0       0  44.0   $44k-$99k     5.4       India,In        1999       TRUE
1       1  66.0   $55k-$66k     3.5    New York,Ny        2002       TRUE
2       2  39.0   $77k-$89k     4.0    New York,Ny     Unknown      FALSE
3       3  64.0   $44k-$99k     4.4       India In        1988      FALSE
4       4  25.0   $44k-$99k     6.4  Australia Aus        2002      FALSE
5       5  44.0   $77k-$89k     1.4       India,In        1999       TRUE
6       6  21.0   $44k-$99k     0.0    New York,Ny     Unknown      FALSE
7       7  44.0   $44k-$99k     4.0  Australia Aus     Unknown      FALSE
8       8  35.0   $44k-$99k     5.4    New York,Ny     Unknown      FALSE
9       9  22.0   $44k-$99k     7.7       India,In     Unknown       TRUE
10     10  55.0   $10k-$49k     5.4       India,In        2008       TRUE
11     11  44.0   $10k-$49k     6.7       India,In        2009      FALSE
12     12  39.0   $44k-$99k     0.0       India,In        1999      FALSE
13     13  25.0   $44k-$99k     4.0  Australia Aus        2019       TRUE
14     14  66.0   $44k-$99k     4.0  Australia Aus        2020       TRUE
15     15  44.0  $88k-$101k     3.0  Australia Aus        1999      FALSE
16     16  19.0   $19k-$40k     4.5       India,In        1984      FALSE
17     17  39.0   $44k-$99k     5.3    New York,Ny        1943       TRUE
18     18  35.0   $44k-$99k     6.7    New York,Ny        1954       TRUE
19     19  32.0   $44k-$99k     3.3    New York,Ny        1955       TRUE
20     20  39.0   $44k-$99k     5.7    New York,Ny        1944       TRUE
21     21  35.0   $44k-$99k     5.0    New York,Ny        1946      FALSE
22     22  19.0   $55k-$66k     7.8    New York,Ny        1988       TRUE
23     23  39.0   $44k-$99k     2.4    New York,Ny        1999       TRUE
24     24  13.0   $44k-$99k     4.0    New York,Ny        1987      FALSE
25     25  55.0   $44k-$99k     0.0  Australia Aus        1980       TRUE
26     26  39.0   $55k-$66k     4.0       India,In        1934       TRUE
27     27  52.0   $44k-$99k     5.4       India,In        1935      FALSE
28     28  39.0   $39k-$88k     3.4  Australia Aus        1932      FALSE

Spliting the `"Salary"` column at `'-'`, forming two columns namely `"Starting Salary"` and `"Ending Salary"`

In [23]:
df[["Starting Salary","Ending Salary"]]=df["Salary"].str.split('-',1,expand=True)
df=df.drop(columns=["Salary"])
df

<ipython-input-23-b7b33b2beec4>:1: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.
  df[["Starting Salary","Ending Salary"]]=df["Salary"].str.split('-',1,expand=True)


Index   Age  Rating       Location Established Easy Apply Starting Salary  \
0       0  44.0     5.4       India,In        1999       TRUE            $44k   
1       1  66.0     3.5    New York,Ny        2002       TRUE            $55k   
2       2  39.0     4.0    New York,Ny     Unknown      FALSE            $77k   
3       3  64.0     4.4       India In        1988      FALSE            $44k   
4       4  25.0     6.4  Australia Aus        2002      FALSE            $44k   
5       5  44.0     1.4       India,In        1999       TRUE            $77k   
6       6  21.0     0.0    New York,Ny     Unknown      FALSE            $44k   
7       7  44.0     4.0  Australia Aus     Unknown      FALSE            $44k   
8       8  35.0     5.4    New York,Ny     Unknown      FALSE            $44k   
9       9  22.0     7.7       India,In     Unknown       TRUE            $44k   
10     10  55.0     5.4       India,In        2008       TRUE            $10k   
11     11  44.0     6.7       India,In        2009      FALSE            $10k   
12     12  39.0     0.0       India,In        1999      FALSE            $44k   
13     13  25.0     4.0  Australia Aus        2019       TRUE            $44k   
14     14  66.0     4.0  Australia Aus        2020       TRUE            $44k   
15     15  44.0     3.0  Australia Aus        1999      FALSE            $88k   
16     16  19.0     4.5       India,In        1984      FALSE            $19k   
17     17  39.0     5.3    New York,Ny        1943       TRUE            $44k   
18     18  35.0     6.7    New York,Ny        1954       TRUE            $44k   
19     19  32.0     3.3    New York,Ny        1955       TRUE            $44k   
20     20  39.0     5.7    New York,Ny        1944       TRUE            $44k   
21     21  35.0     5.0    New York,Ny        1946      FALSE            $44k   
22     22  19.0     7.8    New York,Ny        1988       TRUE            $55k   
23     23  39.0     2.4    New York,Ny        1999       TRUE            $44k   
24     24  13.0     4.0    New York,Ny        1987      FALSE            $44k   
25     25  55.0     0.0  Australia Aus        1980       TRUE            $44k   
26     26  39.0     4.0       India,In        1934       TRUE            $55k   
27     27  52.0     5.4       India,In        1935      FALSE            $44k   
28     28  39.0     3.4  Australia Aus        1932      FALSE            $39k   

   Ending Salary  
0           $99k  
1           $66k  
2           $89k  
3           $99k  
4           $99k  
5           $89k  
6           $99k  
7           $99k  
8           $99k  
9           $99k  
10          $49k  
11          $49k  
12          $99k  
13          $99k  
14          $99k  
15         $101k  
16          $40k  
17          $99k  
18          $99k  
19          $99k  
20          $99k  
21          $99k  
22          $66k  
23          $99k  
24          $99k  
25          $99k  
26          $66k  
27          $99k  
28          $88k

Left Strip the `'$'` symbol for both of the columns.
<br>Replace `'k'` alphabet to `'000'`

In [24]:
df["Starting Salary"]=df["Starting Salary"].str.strip("$")
df["Ending Salary"]=df["Ending Salary"].str.strip("$")
df["Starting Salary"]=df["Starting Salary"].str.replace('k','000')
df["Ending Salary"]=df["Ending Salary"].str.replace('k','000')
df

Index   Age  Rating       Location Established Easy Apply Starting Salary  \
0       0  44.0     5.4       India,In        1999       TRUE           44000   
1       1  66.0     3.5    New York,Ny        2002       TRUE           55000   
2       2  39.0     4.0    New York,Ny     Unknown      FALSE           77000   
3       3  64.0     4.4       India In        1988      FALSE           44000   
4       4  25.0     6.4  Australia Aus        2002      FALSE           44000   
5       5  44.0     1.4       India,In        1999       TRUE           77000   
6       6  21.0     0.0    New York,Ny     Unknown      FALSE           44000   
7       7  44.0     4.0  Australia Aus     Unknown      FALSE           44000   
8       8  35.0     5.4    New York,Ny     Unknown      FALSE           44000   
9       9  22.0     7.7       India,In     Unknown       TRUE           44000   
10     10  55.0     5.4       India,In        2008       TRUE           10000   
11     11  44.0     6.7       India,In        2009      FALSE           10000   
12     12  39.0     0.0       India,In        1999      FALSE           44000   
13     13  25.0     4.0  Australia Aus        2019       TRUE           44000   
14     14  66.0     4.0  Australia Aus        2020       TRUE           44000   
15     15  44.0     3.0  Australia Aus        1999      FALSE           88000   
16     16  19.0     4.5       India,In        1984      FALSE           19000   
17     17  39.0     5.3    New York,Ny        1943       TRUE           44000   
18     18  35.0     6.7    New York,Ny        1954       TRUE           44000   
19     19  32.0     3.3    New York,Ny        1955       TRUE           44000   
20     20  39.0     5.7    New York,Ny        1944       TRUE           44000   
21     21  35.0     5.0    New York,Ny        1946      FALSE           44000   
22     22  19.0     7.8    New York,Ny        1988       TRUE           55000   
23     23  39.0     2.4    New York,Ny        1999       TRUE           44000   
24     24  13.0     4.0    New York,Ny        1987      FALSE           44000   
25     25  55.0     0.0  Australia Aus        1980       TRUE           44000   
26     26  39.0     4.0       India,In        1934       TRUE           55000   
27     27  52.0     5.4       India,In        1935      FALSE           44000   
28     28  39.0     3.4  Australia Aus        1932      FALSE           39000   

   Ending Salary  
0          99000  
1          66000  
2          89000  
3          99000  
4          99000  
5          89000  
6          99000  
7          99000  
8          99000  
9          99000  
10         49000  
11         49000  
12         99000  
13         99000  
14         99000  
15        101000  
16         40000  
17         99000  
18         99000  
19         99000  
20         99000  
21         99000  
22         66000  
23         99000  
24         99000  
25         99000  
26         66000  
27         99000  
28         88000

Converting all uncertain data types to respective and needed data types for example, float to int for `Establishment` or String to Bool for `Easy Apply` using map.<br>

In [25]:
df["Starting Salary"] = df["Starting Salary"].astype(int)
df["Ending Salary"] = df["Ending Salary"].astype(int)
df["Age"] = df["Age"].astype(int)
df['Easy Apply'] = df['Easy Apply'].map(mapping).fillna(False).astype(bool)
df.reset_index(drop=True, inplace=True)
df

Index  Age  Rating       Location Established  Easy Apply  \
0       0   44     5.4       India,In        1999        True   
1       1   66     3.5    New York,Ny        2002        True   
2       2   39     4.0    New York,Ny     Unknown       False   
3       3   64     4.4       India In        1988       False   
4       4   25     6.4  Australia Aus        2002       False   
5       5   44     1.4       India,In        1999        True   
6       6   21     0.0    New York,Ny     Unknown       False   
7       7   44     4.0  Australia Aus     Unknown       False   
8       8   35     5.4    New York,Ny     Unknown       False   
9       9   22     7.7       India,In     Unknown        True   
10     10   55     5.4       India,In        2008        True   
11     11   44     6.7       India,In        2009       False   
12     12   39     0.0       India,In        1999       False   
13     13   25     4.0  Australia Aus        2019        True   
14     14   66     4.0  Australia Aus        2020        True   
15     15   44     3.0  Australia Aus        1999       False   
16     16   19     4.5       India,In        1984       False   
17     17   39     5.3    New York,Ny        1943        True   
18     18   35     6.7    New York,Ny        1954        True   
19     19   32     3.3    New York,Ny        1955        True   
20     20   39     5.7    New York,Ny        1944        True   
21     21   35     5.0    New York,Ny        1946       False   
22     22   19     7.8    New York,Ny        1988        True   
23     23   39     2.4    New York,Ny        1999        True   
24     24   13     4.0    New York,Ny        1987       False   
25     25   55     0.0  Australia Aus        1980        True   
26     26   39     4.0       India,In        1934        True   
27     27   52     5.4       India,In        1935       False   
28     28   39     3.4  Australia Aus        1932       False   

    Starting Salary  Ending Salary  
0             44000          99000  
1             55000          66000  
2             77000          89000  
3             44000          99000  
4             44000          99000  
5             77000          89000  
6             44000          99000  
7             44000          99000  
8             44000          99000  
9             44000          99000  
10            10000          49000  
11            10000          49000  
12            44000          99000  
13            44000          99000  
14            44000          99000  
15            88000         101000  
16            19000          40000  
17            44000          99000  
18            44000          99000  
19            44000          99000  
20            44000          99000  
21            44000          99000  
22            55000          66000  
23            44000          99000  
24            44000          99000  
25            44000          99000  
26            55000          66000  
27            44000          99000  
28            39000          88000

*  `India,In` : Here we have to split `India` and `In` at `','`
*  `New York,NY` : Here also we need to split `New York` and `Ny` at `','`
*  `Australia Aus` : Its a special case where we need to split `Australia` and `Aus` at `' '` but it shouldn't collide with New York, so we need to place them priortywise.


After splitting them we will be alloting them to seperate columns namely `Country` and `Initials`.

After converting, we find that there was incorrect indexing due to dropping of data which is now fixed.

In [26]:
# Splitting based on different separators for different locations
def split_location(row):
    location = row['Location']
    if ',' in location:
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator

    return pd.Series({'Country': country, 'Initials': initials})

# Apply the function to split 'Location' into 'Country' and 'Initials'
df[['Country', 'Initials']] = df.apply(split_location, axis=1)
df=df.drop(columns=["Location"])
df.reset_index(drop=True, inplace=True)
df.drop(columns=['Index'], inplace=True)
df

Age  Rating Established  Easy Apply  Starting Salary  Ending Salary  \
0    44     5.4        1999        True            44000          99000   
1    66     3.5        2002        True            55000          66000   
2    39     4.0     Unknown       False            77000          89000   
3    64     4.4        1988       False            44000          99000   
4    25     6.4        2002       False            44000          99000   
5    44     1.4        1999        True            77000          89000   
6    21     0.0     Unknown       False            44000          99000   
7    44     4.0     Unknown       False            44000          99000   
8    35     5.4     Unknown       False            44000          99000   
9    22     7.7     Unknown        True            44000          99000   
10   55     5.4        2008        True            10000          49000   
11   44     6.7        2009       False            10000          49000   
12   39     0.0        1999       False            44000          99000   
13   25     4.0        2019        True            44000          99000   
14   66     4.0        2020        True            44000          99000   
15   44     3.0        1999       False            88000         101000   
16   19     4.5        1984       False            19000          40000   
17   39     5.3        1943        True            44000          99000   
18   35     6.7        1954        True            44000          99000   
19   32     3.3        1955        True            44000          99000   
20   39     5.7        1944        True            44000          99000   
21   35     5.0        1946       False            44000          99000   
22   19     7.8        1988        True            55000          66000   
23   39     2.4        1999        True            44000          99000   
24   13     4.0        1987       False            44000          99000   
25   55     0.0        1980        True            44000          99000   
26   39     4.0        1934        True            55000          66000   
27   52     5.4        1935       False            44000          99000   
28   39     3.4        1932       False            39000          88000   

      Country Initials  
0       India       In  
1    New York       Ny  
2    New York       Ny  
3       India       In  
4   Australia      Aus  
5       India       In  
6    New York       Ny  
7   Australia      Aus  
8    New York       Ny  
9       India       In  
10      India       In  
11      India       In  
12      India       In  
13  Australia      Aus  
14  Australia      Aus  
15  Australia      Aus  
16      India       In  
17   New York       Ny  
18   New York       Ny  
19   New York       Ny  
20   New York       Ny  
21   New York       Ny  
22   New York       Ny  
23   New York       Ny  
24   New York       Ny  
25  Australia      Aus  
26      India       In  
27      India       In  
28  Australia      Aus

#Code

In [27]:
import pandas as pd  # Importing pandas library
import numpy as np   # Importing numpy library

# Function to split 'Location' into 'Country' and 'Initials'
def split_location(row):
    location = row['Location']
    if ',' in location:  # Split at ','
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:  # Elif Split at ' '
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:  # Else split none
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator

    return pd.Series({'Country': country, 'Initials': initials})  # Assigning each series with the values of a particular row

# Read the CSV file into a DataFrame
df = pd.read_csv('/content/drive/MyDrive/PrepInsta/PrepInsta-DA-Week-3-main/Assignment1.csv')

# Drop duplicate rows
df = df.drop_duplicates()

# Replace '-1' with 'FALSE' in the 'Easy Apply' column
df["Easy Apply"] = df["Easy Apply"].str.replace('-1', 'FALSE')

# Mapping string values to boolean values in the 'Easy Apply' column
mapping = {'TRUE': True, 'FALSE': False}
df['Easy Apply'] = df['Easy Apply'].map(mapping).fillna(False).astype(bool)

# Convert 'Established' to string data type
df["Established"] = df["Established"].astype(str)

# Replace remaining -1 with NaN value
df = df.replace(-1, np.nan)

# Calculate and fill missing values in 'Age' with the mean
age_mean = df["Age"].mean()
age_mean = int(age_mean)
df["Age"] = df["Age"].replace(np.nan, age_mean)

# Calculate and fill missing values in 'Rating' with the mean
rating_mean = df["Rating"].mean()
rating_mean = int(rating_mean)
df["Rating"] = df["Rating"].replace(np.nan, rating_mean)

# Replace '-1' with "Unknown" in the 'Established


In [28]:
# Convert column names to lowercase and replace spaces with underscores
df.columns = df.columns.str.lower().str.replace(' ', '_')

# Display the DataFrame with modified column names
df


index   age      salary  rating       location established  easy_apply
0       0  44.0   $44k-$99k     5.4       India,In        1999        True
1       1  66.0   $55k-$66k     3.5    New York,Ny        2002        True
2       2  39.0   $77k-$89k     4.0    New York,Ny          -1       False
3       3  64.0   $44k-$99k     4.4       India In        1988       False
4       4  25.0   $44k-$99k     6.4  Australia Aus        2002       False
5       5  44.0   $77k-$89k     1.4       India,In        1999        True
6       6  21.0   $44k-$99k     0.0    New York,Ny          -1       False
7       7  44.0   $44k-$99k     4.0  Australia Aus          -1       False
8       8  35.0   $44k-$99k     5.4    New York,Ny          -1       False
9       9  22.0   $44k-$99k     7.7       India,In          -1        True
10     10  55.0   $10k-$49k     5.4       India,In        2008        True
11     11  44.0   $10k-$49k     6.7       India,In        2009       False
12     12  39.0   $44k-$99k     0.0       India,In        1999       False
13     13  25.0   $44k-$99k     4.0  Australia Aus        2019        True
14     14  66.0   $44k-$99k     4.0  Australia Aus        2020        True
15     15  44.0  $88k-$101k     3.0  Australia Aus        1999       False
16     16  19.0   $19k-$40k     4.5       India,In        1984       False
17     17  39.0   $44k-$99k     5.3    New York,Ny        1943        True
18     18  35.0   $44k-$99k     6.7    New York,Ny        1954        True
19     19  32.0   $44k-$99k     3.3    New York,Ny        1955        True
20     20  39.0   $44k-$99k     5.7    New York,Ny        1944        True
21     21  35.0   $44k-$99k     5.0    New York,Ny        1946       False
22     22  19.0   $55k-$66k     7.8    New York,Ny        1988        True
23     23  39.0   $44k-$99k     2.4    New York,Ny        1999        True
24     24  13.0   $44k-$99k     4.0    New York,Ny        1987       False
25     25  55.0   $44k-$99k     0.0  Australia Aus        1980        True
26     26  39.0   $55k-$66k     4.0       India,In        1934        True
27     27  52.0   $44k-$99k     5.4       India,In        1935       False
28     28  39.0   $39k-$88k     3.4  Australia Aus        1932       False

In [29]:
# Display information about the DataFrame including the data types, non-null counts, and memory usage.
df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 29 entries, 0 to 28
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   index        29 non-null     int64  
 1   age          29 non-null     float64
 2   salary       29 non-null     object 
 3   rating       29 non-null     float64
 4   location     29 non-null     object 
 5   established  29 non-null     object 
 6   easy_apply   29 non-null     bool   
dtypes: bool(1), float64(2), int64(1), object(3)
memory usage: 1.6+ KB


#**Problem Solutions:**



###1. Missing Values:
**Question:** <br>Are there any missing values in the dataset, and if so, how should they be handled for each indicator? <br>
**Answer:**<br>
*  Yes, There are missing values in dataset. There are missing values in multiple columns (e.g., `Age, Rating, Established, Easy Apply`).
*  Handling: These missing values need to be addressed, either through imputation, removal of rows, or depending on the context. <br>
*  However column named `"Easy Apply"` has `-1` which means `False`
*  Whereas for columns `"Age"`, `"Rating"`, `"Established"`, we need to replace with mean of Age, mean of Rating and `"Unknown"` values respectively.

The solution to clean missing values:


```
df["Easy Apply"]=df["Easy Apply"].str.replace('-1','FALSE')       # Replace -1 with False in "Easy Apply Column"

age_mean=df["Age"].mean()                                         # For Replacing NaN values with mean of Age
age_mean=int(age_mean)
df["Age"] = df["Age"].replace(np.nan, age_mean)

rating_mean=df["Rating"].mean()                                   # For Replacing NaN value with mean of Rating
rating_mean=int(rating_mean)
df["Rating"] = df["Rating"].replace(np.nan, rating_mean)  

df["Established"] = df["Established"].replace('-1', "Unknown")    # Replace -1 with Unknown
```

###2. Data Types:
**Question:** <br>What are the data types of each indicator, and do they align with their expected types (e.g., numerical, categorical)? <br>
**Answer:**<br>
*  `Age:` Mostly numerical (except for some missing values).
*  `Salary:` String format (needs conversion to numerical values).
*  `Rating:` Floating numerical with some missing values represented as -1.
*  `Location:` String format.
*  `Established:` Mostly numerical with some missing values.
*  `Easy Apply:` Boolean (except for missing values).

The solution to find out the data type of each series:

```
df["Indicator/Column_Name"].dtype           # Returns the fdata type of each indicator
```

###3. Outliers:
**Question:** <br>Identify potential outliers in numerical indicators (e.g., Age, Salary, Rating). Should outliers be removed or adjusted?
 <br>
**Answer:**<br>
*  `Age:` Any age that seems unusually high or low compared to the typical range in the dataset could be considered an outlier.<br>*For example: `13` is seem to be low outlier and `66` seems to be high outlier `Age`.*
*  `Salary:` Outliers might include exceptionally high or low salaries compared to the majority of salaries in the dataset.<br>*For example: `10000` is seem to be low outlier and `101000` seems to be high outlier `Salary`.*
*  `Rating:` Values significantly deviating from the general range of ratings could be identified as outliers.<br>*For example: `0` is seem to be low outlier and `7.7` seems to be high outlier `Rating`.*

The solution to find out the outliers, we use the following algorithm:

```
    q1 = data.quantile(0.25)
    q3 = data.quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    outliers = data[(data < lower_bound) | (data > upper_bound)]
```

###4. Salary Formatting:
**Question:** <br>Examine the format of the Salary column. Does it require any formatting or standardization for consistent analysis?
 <br>
**Answer:**<br>
Salary values are in the format `"$XXk-$XXk"` and require transformation into numerical values.
*  Split the `"Salary"` column at `'-'`, forming two columns namely `"Starting Salary"` and `"Ending Salary"`
*  Left Strip the `'$'` symbol for both of the columns.
*  Replace `'k'` alphabet to `'000'`

Following the the below snippet, we can clean the `"Salary"` column

```
df[["Starting Salary","Ending Salary"]]=df["Salary"].str.split('-',1,expand=True)       # Splits the following Salary column
df=df.drop(columns=["Salary"])                                        # Delete the Salary column

df["Starting Salary"]=df["Starting Salary"].str.strip("$")            # Left Stripping '$'
df["Ending Salary"]=df["Ending Salary"].str.strip("$")
df["Starting Salary"]=df["Starting Salary"].str.replace('k','000')    # Replace 'k' with '000'
df["Ending Salary"]=df["Ending Salary"].str.replace('k','000')
```

###5. Location Standardization:
**Question:** <br>Check the consistency of location entries. Do they need standardization, and how can this be achieved?
 <br>
**Answer:**<br>
`Locations` have various formats (e.g., "India,In", "New York,Ny", "Australia Aus"). Standardization could involve splitting these into separate columns for `Country` and `Initials`.
Let us take a tour of all the countries,
*  `India,In` : Here we have to split `India` and `In` at `','`
*  `New York,NY` : Here also we need to split `New York` and `Ny` at `','`
*  `Australia Aus` : Its a special case where we need to split `Australia` and `Aus` at `' '` but it shouldn't collide with New York, so we need to place them priortywise.

The following steps show how to split the `Location` into two seperate columns.



```
# Splitting based on different separators for different locations
def split_location(row):
    location = row['Location']
    if ',' in location:
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator

    return pd.Series({'Country': country, 'Initials': initials})

# Apply the function to split 'Location' into 'Country' and 'Initials'
df[['Country', 'Initials']] = df.apply(split_location, axis=1)
df=df.drop(columns=["Location"])
```

###6. Established Column:
**Question:** <br>Explore the Established column. Are there any inconsistencies or anomalies that need to be addressed?
 <br>
**Answer:**<br>
Yes, Inconsistencies exist in the Established column, with missing values represented as -1. It is converted to string first, then replace -1 with "Unknown" string

It can be solved if emphsize the following snippet:



```
df["Established"] = df["Established"].astype(str)       # Converting them to string data type
df["Established"] = df["Established"].replace('-1', "Unknown") # Replace -1 with Unknown
```

###7. Easy Apply Indicator:
**Question:** <br>Analyze the Easy Apply column. Does it contain boolean values or need transformation for better analysis?
 <br>
**Answer:**<br>
`Easy Apply` column initially consists of string data type. Apart from that, it also consists of missing values which is `"-1"`. However, we consider `-1` as `FALSE` and replace all `-1` with `FALSE`.

Still, it is carries, string data type, which is converted to bool using map function.

Following algorithm is used for that case:

```
df["Easy Apply"]=df["Easy Apply"].str.replace('-1','FALSE')                   # Replace -1 with False
mapping = {'TRUE': True, 'FALSE': False}                                      # Mapped the string to respective bool value
df['Easy Apply'] = df['Easy Apply'].map(mapping).fillna(False).astype(bool)   # Formatting the column with Boolean data type
```

###8. Rating Range:
**Question:** <br>Investigate the range of values in the Rating column. Does it fall within expected rating scales, and how should outliers be treated?
<br>
**Answer:**<br>
Ratings range from `0 to 7.8`, with missing values represented as `-1`. It is justified that, it lies on a rating scale of 10 points. Moreover, missing values may need addressing.


For that kind of scenarios, we first replace the missing values with `NaN` then find the mean of `"Rating"` column, then replace all NaN values with its mean:

```
rating_mean=df["Rating"].mean()                                   # For Replacing NaN value with mean of Rating
rating_mean=int(rating_mean)
df["Rating"] = df["Rating"].replace(np.nan, rating_mean)  
```

###9. Age Distribution:
**Question:** <br>Check the distribution of values in the Age column. Are there any unusual entries, and how might they impact analysis?
<br>
**Answer:**<br>
`Age` values are present, but some are missing. Apart from that, we can see certain outliers in the `Age` column. Any age that seems unusually high or low compared to the typical range in the dataset could be considered an outlier.
For example: `13` is seem to be low outlier and `66` seems to be high outlier Age.


In this way we delete rows where `Age` is `NaN`:

```
df=df.dropna()                              # Delete rows consisting NaN in data frame
```

###10. Handling Special Characters:
**Question:** <br>Examine all text-based columns (e.g., Location). Are there special characters or inconsistencies that need cleaning?
<br>
**Answer:**<br>
`Location` entries contain commas and spaces that require splitting or cleaning.
`Locations` have various formats (e.g., "India,In", "New York,Ny", "Australia Aus"). Standardization could involve splitting these into separate columns for `Country` and `Initials`.
Let us take a tour of all the countries,
*  `India,In` : Here we have to split `India` and `In` at `','`
*  `New York,NY` : Here also we need to split `New York` and `Ny` at `','`
*  `Australia Aus` : Its a special case where we need to split `Australia` and `Aus` at `' '` but it shouldn't collide with New York, so we need to place them priortywise.

The following steps show how to split the `Location` into two seperate columns.


```
# Splitting based on different separators for different locations
def split_location(row):
    location = row['Location']
    if ',' in location:
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator

    return pd.Series({'Country': country, 'Initials': initials})

# Apply the function to split 'Location' into 'Country' and 'Initials'
df[['Country', 'Initials']] = df.apply(split_location, axis=1)
df=df.drop(columns=["Location"])
```

`Salary values` are in the format `"$XXk-$XXk"` and require transformation into numerical values.
*  Split the `"Salary"` column at `'-'`, forming two columns namely `"Starting Salary"` and `"Ending Salary"`
*  Left Strip the `'$'` symbol for both of the columns.
*  Replace `'k'` alphabet to `'000'`

Following the the below snippet, we can clean the `"Salary"` column

```
df[["Starting Salary","Ending Salary"]]=df["Salary"].str.split('-',1,expand=True)       # Splits the following Salary column
df=df.drop(columns=["Salary"])                                        # Delete the Salary column

df["Starting Salary"]=df["Starting Salary"].str.strip("$")            # Left Stripping '$'
df["Ending Salary"]=df["Ending Salary"].str.strip("$")
df["Starting Salary"]=df["Starting Salary"].str.replace('k','000')    # Replace 'k' with '000'
df["Ending Salary"]=df["Ending Salary"].str.replace('k','000')
```

###11. Data Integrity:
**Question:** <br>Ensure data integrity by cross-referencing entries. For instance, does the Established column align with the Age column?
<br>
**Answer:**<br>
Yes, Inconsistencies exist between the `Established` and `Age` columns, with missing values causing misalignment.

*  The `Established` column denotes the year of establishment for certain records, represented mostly in numerical format but contains missing values indicated by -1.
*  The `Age` column refers to the age of individuals or entities, but it also contains missing values.
*  We can compare the `Established` column's values (years) with the `Age` column to identify inconsistencies or misalignments. If both columns refer to the same entity, the established year should correlate with the age or else not.



###12. Easy Apply Transformation:
**Question:** <br>If the Easy Apply column contains non-boolean values, how can it be transformed into a usable format?
<br>
**Answer:**<br>
Some missing values exist in the `Easy Apply` column that may require handling.
`Easy Apply` column initially consists of string data type. Apart from that, it also consists of missing values which is `"-1"`. However, we consider `-1` as `FALSE` and replace all `-1` with `FALSE`.

Still, it is carries, string data type, which is converted to bool using map function.

Following algorithm is used for that case:

```
df["Easy Apply"]=df["Easy Apply"].str.replace('-1','FALSE')                   # Replace -1 with False
mapping = {'TRUE': True, 'FALSE': False}                                      # Mapped the string to respective bool value
df['Easy Apply'] = df['Easy Apply'].map(mapping).fillna(False).astype(bool)   # Formatting the column with Boolean data type
```



###13. Location Accuracy:
**Question:** <br>Assess the accuracy of location entries. Are there misspelled or ambiguous locations that require correction?
<br>
**Answer:**<br>
Entries like `"India,In"` and `"New York,Ny"` could use splitting into separate columns for country and initials at `','`.
<br> Apart from that `"Australia Aus"` could be use splitting into the same columns at `' '`.
Also, we need to look out for that if we split at `' '` first then, `New York` would be affected. To avoid it we can use them in `if elif` contional flow, where `','` comes priorly than `' '`.

```
if ',' in location:
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:               # Specially for Australia
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator
```



###14. Handling Categorical Data:
**Question:** <br>For categorical indicators, consider encoding or transforming them into a format suitable for analysis.
<br>
**Answer:**<br>
Handling categorical data in the `"Location"` column involves:

*  **Standardization**: Splitting the column into `"Country"` and `"Initials"` for uniformity and consistency.
*  **Cleaning**: Removing special characters and maintaining consistent formatting for locations.
*  **Transformation**: Encoding categorical data into numerical representations for analysis, using methods like label encoding.

```
if ',' in location:
        parts = location.split(',')
        country = parts[0].strip()
        initials = parts[1].strip()
    elif ' ' in location:               # Specially for Australia
        parts = location.split(' ')
        country = parts[0].strip()
        initials = parts[1].strip()
    else:
        country = location.strip()
        initials = None  # Or set initials as needed for cases with no separator
```



###15. Consistent Rating Scale:
**Question:** <br>Ensure a consistent rating scale in the Rating column. Should it be normalized or adjusted for uniform analysis?
<br>
**Answer:**<br>
Outliers might include exceptionally high or low salaries compared to the majority of salaries in the dataset.<br>*For example: `10000` is seem to be low outlier and `101000` seems to be high outlier `Salary`.*
<br>Ratings range from `0 to 7.8`, with missing values represented as `-1`. It is justified that, it lies on a rating scale of 10 points. Moreover, missing values may need addressing.


For that kind of scenarios, we first replace the missing values with `NaN` then find the mean of `"Rating"` column, then replace all NaN values with its mean:

```
rating_mean=df["Rating"].mean()                                   # For Replacing NaN value with mean of Rating
rating_mean=int(rating_mean)
df["Rating"] = df["Rating"].replace(np.nan, rating_mean)  
```

